File: python-sdk/src/abapay/x402.py
"""Zero-setup x402 payments — mirrors sdk/src/x402.ts's payBillViaX402 field-for-field. The
agent's own wallet signs an EIP-3009 transferWithAuthorization for exactly the amount AbaPay's
server names in its 402 challenge; nothing here ever touches abapays.com, an Agent Hub
account, or a PIN. Celo mainnet only in this version.

Verify this against the real wire format any time: https://agents.abapays.com/x402
"""

from __future__ import annotations

import base64
import json
import os
import time
import urllib.error
import urllib.request
from typing import Any, Optional

from eth_account.messages import encode_typed_data

from .types import AbaPayError, BillDetails, X402PayResult

DEFAULT_BASE_URL = "https://www.abapays.com"

# ⚡ MUST MATCH src/lib/x402Settle.ts's TRANSFER_WITH_AUTHORIZATION_TYPES BYTE-FOR-BYTE (and
# sdk/src/x402.ts's copy of the same) — that's the code that reconstructs this signature
# server-side to verify it. Three independent copies now (TypeScript app, TypeScript SDK,
# this one) because none of them can depend on each other across language/publish boundaries;
# a drift in any one is a signature that silently fails to verify. Each carries this exact
# comment as the tripwire.
TRANSFER_WITH_AUTHORIZATION_TYPES = {
    "TransferWithAuthorization": [
        {"name": "from", "type": "address"},
        {"name": "to", "type": "address"},
        {"name": "value", "type": "uint256"},
        {"name": "validAfter", "type": "uint256"},
        {"name": "validBefore", "type": "uint256"},
        {"name": "nonce", "type": "bytes32"},
    ],
}


def _random_nonce() -> str:
    return "0x" + os.urandom(32).hex()


def _caip2_chain_id(network: str) -> int:
    parts = network.split(":")
    if len(parts) != 2 or parts[0] != "eip155":
        raise AbaPayError(
            f'Unsupported x402 network "{network}" — this SDK version only understands '
            "eip155:* (EVM) networks, and only Celo (eip155:42220) is currently offered by AbaPay."
        )
    try:
        return int(parts[1])
    except ValueError:
        raise AbaPayError(f'Unsupported x402 network "{network}" — non-numeric chain id.')


def _post_json(url: str, body: dict, extra_headers: Optional[dict] = None) -> tuple[int, dict]:
    headers = {"Content-Type": "application/json"}
    if extra_headers:
        headers.update(extra_headers)
    data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(url, data=data, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req) as res:
            raw = res.read().decode("utf-8")
            return res.status, json.loads(raw)
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8")
        try:
            return e.code, json.loads(raw)
        except json.JSONDecodeError:
            raise AbaPayError(f"AbaPay returned a non-JSON response (HTTP {e.code}): {raw[:300]}")


def pay_bill_via_x402(
    signer: Any,
    bill: BillDetails,
    base_url: str = DEFAULT_BASE_URL,
) -> X402PayResult:
    """Pay a real-world bill (airtime, data, electricity, cable) via x402 — zero setup.

    `signer` is any eth_account LocalAccount (e.g. from
    `eth_account.Account.from_key(private_key)`). Its private key never leaves this process —
    only `sign_message`/typed-data signing is used, matching the SDK's own no-custody design.

    Raises AbaPayError on any HTTP failure, a challenge this version can't act on, or a final
    response AbaPay itself reports as unsuccessful.
    """
    if bill.wallet_address is None:
        bill = BillDetails(**{**bill.__dict__, "wallet_address": signer.address})
    endpoint = f"{base_url.rstrip('/')}/api/pay/x402"

    # ── 1. Ask for a price. No X-PAYMENT header yet — this MUST come back 402. ──────────────
    status, challenge = _post_json(endpoint, bill.to_dict())
    if status != 402:
        raise AbaPayError(f"Expected HTTP 402 (payment required), got {status}.", response=challenge)

    accepts = challenge.get("accepts") or []
    accept = next((a for a in accepts if a.get("scheme") == "exact"), None)
    if accept is None:
        raise AbaPayError("402 challenge carried no usable 'exact' payment option.", response=challenge)

    # ── 2. Sign the exact amount the challenge named, with the wallet's OWN key. ─────────────
    chain_id = _caip2_chain_id(accept["network"])
    now = int(time.time())
    authorization = {
        "from": signer.address,
        "to": accept["payTo"],
        "value": int(accept["amount"]),
        # 60s of slack behind "now" absorbs ordinary clock skew.
        "validAfter": now - 60,
        "validBefore": now + int(accept.get("maxTimeoutSeconds") or 3600),
        "nonce": _random_nonce(),
    }

    domain = {
        "name": accept["extra"]["name"],
        "version": accept["extra"]["version"],
        "chainId": chain_id,
        "verifyingContract": accept["asset"],
    }
    signable = encode_typed_data(
        domain_data=domain,
        message_types=TRANSFER_WITH_AUTHORIZATION_TYPES,
        message_data=authorization,
    )
    signed = signer.sign_message(signable)
    signature = signed.signature.hex()
    if not signature.startswith("0x"):
        signature = "0x" + signature

    # ── 3. Retry with the signed authorization attached. ─────────────────────────────────────
    payment_header = base64.b64encode(
        json.dumps(
            {
                "x402Version": 1,
                "scheme": "exact",
                "network": "celo",
                "payload": {
                    "signature": signature,
                    "authorization": {
                        **authorization,
                        "value": str(authorization["value"]),
                        "validAfter": str(authorization["validAfter"]),
                        "validBefore": str(authorization["validBefore"]),
                    },
                },
            }
        ).encode("utf-8")
    ).decode("ascii")

    settle_status, result = _post_json(endpoint, bill.to_dict(), {"X-PAYMENT": payment_header})
    parsed = X402PayResult.from_dict(result)
    if settle_status >= 300 or parsed.success is False:
        raise AbaPayError(parsed.message or f"Settlement failed (HTTP {settle_status}).", response=result)
    return parsed

File: sdk/src/x402.ts
import type { Hex } from "viem";
import type { BillDetails, Signer, X402PayResult } from "./types.js";
import { AbaPayError } from "./types.js";

export const DEFAULT_BASE_URL = "https://agents.abapays.com";

// ⚡ THE EIP-712 TYPES — must match src/lib/x402Settle.ts's TRANSFER_WITH_AUTHORIZATION_TYPES
// in the main AbaPay repo byte-for-byte, since that file is what verifies the signature this
// produces. Duplicated here (rather than imported) because this package is published
// standalone and cannot depend on the Next.js app's internal src/lib — if AbaPay ever changes
// this type, the change has to land in both places, and this comment is the tripwire for that.
const TRANSFER_WITH_AUTHORIZATION_TYPES = {
  TransferWithAuthorization: [
    { name: "from", type: "address" },
    { name: "to", type: "address" },
    { name: "value", type: "uint256" },
    { name: "validAfter", type: "uint256" },
    { name: "validBefore", type: "uint256" },
    { name: "nonce", type: "bytes32" },
  ],
} as const;

interface X402Accept {
  scheme: string;
  network: string; // CAIP-2, e.g. "eip155:42220"
  amount: string;
  maxAmountRequired: string;
  payTo: string;
  asset: string;
  maxTimeoutSeconds: number;
  extra: { name: string; version: string; primaryType: string };
}

interface X402Challenge {
  x402Version: number;
  error: string;
  accepts: X402Accept[];
}

// Runtime-agnostic base64 — this SDK has no hard Node dependency (agents run in browsers,
// edge functions, and Deno/Bun too), so `Buffer` isn't assumed available.
function toBase64(json: string): string {
  if (typeof Buffer !== "undefined") return Buffer.from(json, "utf-8").toString("base64");
  const bytes = new TextEncoder().encode(json);
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary);
}

function randomNonce(): Hex {
  const bytes = new Uint8Array(32);
  // Prefer Web Crypto (Node 18+, browsers, edge runtimes) — no extra dependency, works
  // everywhere this package's peer (viem) already assumes a modern runtime.
  if (typeof crypto !== "undefined" && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < bytes.length; i++) bytes[i] = Math.floor(Math.random() * 256);
  }
  return `0x${Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("")}` as Hex;
}

function caip2ChainId(network: string): number {
  const parts = network.split(":");
  const id = Number(parts[1]);
  if (parts[0] !== "eip155" || !Number.isFinite(id)) {
    throw new AbaPayError(
      `Unsupported x402 network "${network}" — this SDK version only understands eip155:* (EVM) networks, and only Celo (eip155:42220) is currently offered by AbaPay.`,
    );
  }
  return id;
}

async function readJson(res: Response): Promise<any> {
  const text = await res.text();
  try {
    return JSON.parse(text);
  } catch {
    throw new AbaPayError(`AbaPay returned a non-JSON response (HTTP ${res.status}): ${text.slice(0, 300)}`);
  }
}

/**
 * Pay a real-world bill (airtime, data, electricity, cable) via x402 — zero setup.
 *
 * The agent's own wallet signs an EIP-3009 `transferWithAuthorization` for exactly the amount
 * AbaPay's server names in its 402 challenge; nothing here ever touches abapays.com, an Agent
 * Hub account, or a PIN. This is the same flow documented at https://agents.abapays.com/#x402 —
 * see there for the full request/response shapes this function builds.
 *
 * Celo mainnet only in this version. `bill.token` must be a token the challenge actually offers
 * (USDC or USDT on Celo, both EIP-3009-capable) — offering an unsupported token surfaces as the
 * `AbaPayError` thrown when the challenge has no matching `accepts[]` entry, not a silent
 * fallback to a token you didn't ask for.
 *
 * @throws {AbaPayError} on any HTTP failure, a challenge this SDK version can't act on, or a
 *   final response AbaPay itself reports as unsuccessful (`success: false`).
 */
export async function payBillViaX402(params: {
  signer: Signer;
  bill: Omit<BillDetails, "wallet_address"> & { wallet_address?: string };
  baseUrl?: string;
}): Promise<X402PayResult> {
  const { signer, baseUrl = DEFAULT_BASE_URL } = params;
  const bill: BillDetails = { ...params.bill, wallet_address: params.bill.wallet_address ?? signer.address };
  const endpoint = `${baseUrl.replace(/\/$/, "")}/api/pay/x402`;

  // ── 1. Ask for a price. No X-PAYMENT header yet — this MUST come back 402. ──────────────
  const challengeRes = await fetch(endpoint, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(bill),
  });

  if (challengeRes.status !== 402) {
    // A well-behaved x402 resource never settles a payment it hasn't been paid for — a
    // non-402 here (200, 4xx, 5xx) means something is wrong with the request itself, not a
    // reason to sign anything.
    const body = await readJson(challengeRes).catch(() => null);
    throw new AbaPayError(
      `Expected HTTP 402 (payment required), got ${challengeRes.status}.`,
      undefined,
      body,
    );
  }

  const challenge: X402Challenge = await readJson(challengeRes);
  const accept = challenge.accepts?.find((a) => a.scheme === "exact");
  if (!accept) {
    throw new AbaPayError("402 challenge carried no usable 'exact' payment option.", undefined, challenge);
  }

  // ── 2. Sign the exact amount the challenge named, with the wallet's OWN key. ─────────────
  const chainId = caip2ChainId(accept.network);
  const now = Math.floor(Date.now() / 1000);
  const authorization = {
    from: signer.address,
    to: accept.payTo,
    value: accept.amount,
    // 60s of slack behind "now" absorbs ordinary clock skew between this machine and
    // whichever node eventually checks validAfter on-chain.
    validAfter: String(now - 60),
    validBefore: String(now + (accept.maxTimeoutSeconds || 3600)),
    nonce: randomNonce(),
  };

  const signature = await signer.signTypedData({
    domain: {
      name: accept.extra.name,
      version: accept.extra.version,
      chainId,
      verifyingContract: accept.asset as Hex,
    },
    types: TRANSFER_WITH_AUTHORIZATION_TYPES,
    primaryType: "TransferWithAuthorization",
    message: {
      from: authorization.from as Hex,
      to: authorization.to as Hex,
      value: BigInt(authorization.value),
      validAfter: BigInt(authorization.validAfter),
      validBefore: BigInt(authorization.validBefore),
      nonce: authorization.nonce,
    },
  });

  // ── 3. Retry with the signed authorization attached. ─────────────────────────────────────
  // x402Version/scheme/network in this envelope are metadata AbaPay's server does not trust
  // from the client and overrides with the one combo proven to settle on its facilitator — see
  // src/app/api/pay/x402/route.ts's own comment on this exact point. Sending them is required
  // by the wire format; their value here doesn't affect what actually settles.
  const paymentHeader = toBase64(
    JSON.stringify({
      x402Version: 1,
      scheme: "exact",
      network: "celo",
      payload: { signature, authorization },
    }),
  );

  const settleRes = await fetch(endpoint, {
    method: "POST",
    headers: { "Content-Type": "application/json", "X-PAYMENT": paymentHeader },
    body: JSON.stringify(bill),
  });
  const result: X402PayResult = await readJson(settleRes);

  if (!settleRes.ok || result.success === false) {
    throw new AbaPayError(result.message || `Settlement failed (HTTP ${settleRes.status}).`, undefined, result);
  }
  return result;
}
